realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/SyncedRealmMigrationTests.kt

```diff
@@ -0,0 +1,286 @@
+            assertTrue(realm.schema[className]!!.hasField(StringOnly.FIELD_CHARS)) // Field has been added
+        }
+    }
+
+    // Check that the Realm can still be opened even if the ondisk schema has more fields than in the model class.
+    // The underlying field should not be deleted, just hidden.
+    @Test
+    fun missingFields_hiddenSilently() {
+        val config = configFactory.createSyncConfigurationBuilder(createTestUser(app))
+                .schema(StringOnly::class.java)
+                .build()
+
+        // Setup initial Realm schema (with too many fields)
+        val className = StringOnly::class.java.simpleName
+        DynamicRealm.getInstance(config).use { dynamicRealm ->
+            val schema = dynamicRealm.schema
+            dynamicRealm.executeTransaction {
+                schema.create(className)
+                        .addField(StringOnly.FIELD_CHARS, String::class.java)
+                        .addField("newField", String::class.java)
+                // A schema version has to be set otherwise Object Store will try to initialize the schema again and reach an
+                // error branch. That is not a real case.
+                dynamicRealm.version = 0
+            }
+        }
+
+        // Open typed Realm, which will validate the schema
+        Realm.getInstance(config).use { realm ->
+            val stringOnlySchema = realm.schema[className]!!
+            assertTrue(stringOnlySchema.hasField(StringOnly.FIELD_CHARS))
+            assertTrue(stringOnlySchema.hasField("newField"))
+            assertEquals(2, stringOnlySchema.fieldNames.size.toLong())
+        }
+    }
+
+    // Check that a Realm cannot be opened if it contain breaking schema changes, like changing a primary key
+    @Test
+    fun breakingSchemaChange_throws() {
+        val config = configFactory.createSyncConfigurationBuilder(createTestUser(app))
+                .schema(PrimaryKeyAsString::class.java)
+                .build()
+
+        // Setup initial Realm schema (with a different primary key)
+        val expectedObjectSchema = OsObjectSchemaInfo.Builder(PrimaryKeyAsString.CLASS_NAME, 2, 0)
+                .addPersistedProperty(PrimaryKeyAsString.FIELD_PRIMARY_KEY, RealmFieldType.STRING, false, true, false)
+                .addPersistedProperty(PrimaryKeyAsString.FIELD_ID, RealmFieldType.INTEGER, true, true, true)
+                .build()
+        val list: MutableList<OsObjectSchemaInfo> = ArrayList()
+        list.add(expectedObjectSchema)
+        val schemaInfo = OsSchemaInfo(list)
+        val configBuilder = OsRealmConfig.Builder(config).schemaInfo(schemaInfo)
+        OsSharedRealm.getInstance(configBuilder, OsSharedRealm.VersionID.LIVE).close()
+        thrown.expectMessage(
+                CoreMatchers.containsString("The following changes cannot be made in additive-only schema mode:"))
+        thrown.expect(IllegalStateException::class.java)
+        Realm.getInstance(config)
```

Comment:
Instead of the `ExpectedException` rule, perhaps we should create our own variant of `assetFailsWith` that takes the expected error string, so you can do something like

```
assertFailsWith<IllegalStateException)("The following changes cannot be made in additive-only schema mode:") {
  Realm.getInstance(config)
}
```

Reply:
Yep, see details on ealier comment.

Reply:
https://github.com/realm/realm-java/pull/6861/files#diff-c94fbae13da3fa85fe77b8c175bbe010R134